### PR TITLE
Garbage Collect completed/failed operations periodically 

### PR DESCRIPTION
**Pull request by chetan-rns**
#### Description:
- Added a new field `gc_expiration_time` that indicates the amount of time in seconds to wait before removing the operation from the DB.
- Added a goroutine that continuously garbage collects completed/failed operations periodically.
- Unit tests 
#### Link to JIRA Story (if applicable):

https://issues.redhat.com/browse/GITOPSRVCE-174

### CODE REVIEW

File: cluster-agent/controllers/managed-gitops/operation_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2021.
```
Comment (drpaneas) [inline cluster-agent/controllers/managed-gitops/operation_controller_test.go:2]:
```suggestion
Copyright 2022.
```

**Review by jparsai — APPROVED** [review]

File: cluster-agent/controllers/managed-gitops/operation_controller.go [hunk 2]
```diff
@@ -60,3 +70,70 @@ func (r *OperationReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+					Name:      fmt.Sprintf("operation-%s", operation.Operation_id),
+					Namespace: util.GetGitOpsEngineSingleInstanceNamespace(),
+				},
+			}
+			err = g.k8sClient.Delete(ctx, operationCR)
+			if err != nil {
+				log.Error(err, "failed to delete operation from the cluster", "operation_id", operation.Operation_id)
+			}
+
+			log.Info("successfully garbage collected operation", "operation_id", operation.Operation_id)
```
Comment (drpaneas) [inline cluster-agent/controllers/managed-gitops/operation_controller.go:136]: If the operation CR fails to be deleted from the cluster, you will end up with two logs. One log msg saying there is an error and another one saying that the operation has been garbage collected. To me, a garbage collected operation means that's deleted from both places (db + cluster).

```suggestion
			err = g.k8sClient.Delete(ctx, operationCR)
			if err != nil {
				log.Error(err, "failed to delete operation from the cluster", "operation_id", operation.Operation_id)
			} else {
				log.Info("successfully garbage collected operation", "operation_id", operation.Operation_id)
			}
```

What happens to the Operation that is failed to be deleted from the cluster but it has successfully been deleted from the database?

**Review by drpaneas — COMMENTED** [review]
@chetan-rns great work here, very nice clean code and I was very happy to see the Gomega matcher pattern in use :D I left a minor comment for the unit-test and a question I have for what is the definition of garbage collected operation. For example, a garbage collected operation is considered as such when it's been deleted from both places (cluster + database) ? or just from the database? And a follow-up question would be, what happens to the the Operation CR that is deleted from the database but still presented on the cluster?

One proposal for this ^ scenario: If the deletion from the database succeeds, but the deletion from the cluster fails, then update the CR on the cluster with a new state `orphaned`. Then use [Kubernetes CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) to run periodically on a given schedule and delete these orphaned operations.

File: backend-shared/config/db/operations_test.go [hunk 3]
```diff
@@ -104,4 +120,68 @@ var _ = Describe("Operations Test", func() {
...
+			Resource_id:             "test-fake-resource-id",
+			Resource_type:           "GitopsEngineInstance",
+			Operation_owner_user_id: testClusterUser.Clusteruser_id,
+			Last_state_update:       time.Now(),
+		}
+
+		err := dbq.CreateOperation(ctx, sampleOperation, sampleOperation.Operation_owner_user_id)
+		Expect(err).To(BeNil())
+
+		By("operation in waiting state shouldn't be returned")
```
[the remark this answers is not in the document]
Comment (chetan-rns) [reply backend-shared/config/db/operations_test.go:138]: Agreed. I've added a check

**Comment by chetan-rns** [comment]
I've updated the PR to use `TaskRetryLoop` where we retry until the Operation CR is removed from the cluster.

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great, thanks @chetan-rns! Some minor tweaks then good to go.

File: backend-shared/config/db/operations_test.go [hunk 4]
```diff
@@ -104,4 +120,69 @@ var _ = Describe("Operations Test", func() {
 		Expect(true).To(Equal(db.IsMaxLengthError(err)))
 
 	})
+
+	It("list all operations to be garbage collected", func() {
+		By("create a sample operation")
```
[the remark this answers is not in the document]
Comment (chetan-rns) [reply backend-shared/config/db/operations_test.go:125]: It makes sense to use `It(...)` here for individual cases. I've updated the unit test

**Review by jgwest — APPROVED** [review]
Looks great, thanks again @chetan-rns!

**Pull request merged**